text/1008-tls-and-alpn.rst

```diff
@@ -0,0 +1,551 @@
+``credentials.json`` file for all future connections to the same
+instance::
+
+    edgedb authenticate
+
+    Authenticate to a remote EdgeDB instance and assign an instance name
+    to simplify future connections.
+
+    USAGE:
+        edgedb authenticate [OPTIONS] <host:port>
+
+    ARGS:
+        <host:port> IP/DNS name and the port of the target instance.
+
+    OPTIONS:
+        --name <name>
+            Specify a new instance name for the remote server. If not
+            present, the name will be interactively asked.
+
+        --user <user>
+            The database user to log into the remote server. If not
+            present, the username will be interactively asked.
+
+        --password <password>
+            The password for the database user to log into the remote
+            server. If not present, the username will be interactively
+            asked. This is also available as an environment variable
+            `EDGEDB_PASSWORD`.
+
+        --database <database>
+            The name of the default database to connect to.
+
+For example::
+
+    $ edgedb authenticate db.example.org:5656
+    User: john
+    Password: ******
+    Default database: edgedb
+    Here is the server certificate:
+      Hostname: db.example.org
+      Org: Company Inc.
+      Fingerprints:
+        SHA-256: 63:2B:11:99:44:40:17:DF:37:FC:C3:DF:0F:3D:15
+    Confirm? [Y/n] Y
+    Login successful.
+    Please specify a name for this instance: [db_example_org_5656]
+    Credential file created, you can now connect to the database with:
+        edgedb -I db_example_org_5656
+
+The user is responsible for trusting the server certificate, because
+trusting unknown certificates in production may lead to MITM attacks.
+This command also verifies the user login information with the server
+and only create a corresponding ``credentials.json`` file if the login
+is successful. In the above example,
+``~/.edgedb/credentials/db_example_org_5656.json`` is created::
+
+    {
+        "host": "db.example.org",
+        "port": 5656,
+        "user": "john",
+        "password": "login-password-in-clear-text",
+        "database": "edgedb",
+        "tls_certdata": "-----BEGIN CERTIFICATE-----\nMIICvjCCAaagAw..."
+    }
+
+And then the client logic for server certificate verification is just
+the same as for local development as explained earlier in this section.
+
+
+ALPN and Protocol Changes
+=========================
+
+The ALPN support in target programming languages:
+
+* Python [4]_: ``set_alpn_protocols()`` and ``selected_alpn_protocol()``
+* Go [5]_: ``SupportedProtos`` and ``NegotiatedProtocol``
+* Node.js [6]_: ``ALPNProtocols`` and ``alpnProtocol``
+
+For now, the EdgeDB server will advertise two protocols in ALPN (however
+EdgeDB is not limited to only these two for future possibilities):
+
+* ``edgedb-binary``: The EdgeDB binary protocol
+* ``http``: HTTP-based protocol, including the server system API, and
```

Comment:
Specifically, this would be `http/1.1`, as we might also add support for `h2` in the future too.